### PR TITLE
SCT-285 ElasticIndexingStorage cleanup part 2

### PR DESCRIPTION
**Pull request by arfathpasha**
- updated architecture diagram
- used ImmutableOf whereever possible in ElasticIndexingStorage
- minor fix to dockerfile
- added todo notes

### CODE REVIEW

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java [hunk 1]
```diff
@@ -1780,57 +1655,72 @@ private void createTables(String indexName, List<IndexingRules> indexingRules) t
...
         props.put(OBJ_MD5, keyword);
 
-        tmp = new LinkedHashMap<>();
-        tmp.put("type", "date");
+        tmp = ImmutableMap.of("type", "date");
         props.put("timestamp", tmp);
 
-        tmp = new LinkedHashMap<>();
-        tmp.put("type", "keyword");
+        tmp = ImmutableMap.of("type", "keyword");
```
Comment (MrCreosote) [inline lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: There's already a keyword var for this above, just do props.put("prefix", keyword)

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java [hunk 2]
```diff
@@ -1780,57 +1655,72 @@ private void createTables(String indexName, List<IndexingRules> indexingRules) t
...
         props.put("islast", tmp);
 
-        tmp = new LinkedHashMap<>();
-        tmp.put("type", "boolean");
+        tmp = ImmutableMap.of("type", "boolean");
         props.put("public", tmp);
 
-        tmp = new LinkedHashMap<>();
-        tmp.put("type", "boolean");
+        tmp = ImmutableMap.of("type", "boolean");
```
Comment (MrCreosote) [inline lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: Could do something similar here and make a bool map and reuse it

**Comment by MrCreosote** [comment]
The reason I like ImmutableMap is because it allows you to do this:

foo = ImmutableMap.of(
    "foo", "bar",
    "thing", ImmutableMap.of(
        "a", "b",
        "c", ImmutableMap.of(
            "thing", "otherthing",
            "thing2", "yetanotherthing"
            )
        ),
    "whee", Arrays.asList(ImmutableMap.of(
        "stuff", "morestuff",
        "stuff2", "evenmorestuff"
        ),
    "thingy", "thinger"
    )

Basically it's almost, but not quite, as good as a map literal.

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java [hunk 3]
```diff
@@ -367,29 +363,24 @@ public void flushIndexing(SearchObjectType objectType) throws IOException {
             GUID guid = new GUID((String) obj.get("guid"));
             ret.put(guid, id);
         }
-        return ret;
+        return ImmutableMap.copyOf(ret);
```
[the remark this answers is not in the document]
Comment (arfathpasha) [reply lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: Just an attempt to be as consistent as possible to keep all maps immutable. Some maps are built using if-else logic and I found the only way to ensure their immutability was to first build them and then make them immutable.

[the remark this answers is not in the document]
Comment (MrCreosote) [reply lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: Ah. I don't really worry about immutability of outputs unless modification of the outputs can change internal state.
